src/reconcile/rec_write.c

```diff
@@ -2386,31 +2391,39 @@ __rec_split_write(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REC_CHUNK *chunk
 
     /* Check the eviction flag as checkpoint also saves updates. */
     if (F_ISSET(r, WT_REC_EVICT) && multi->supd != NULL) {
-        /*
-         * XXX If no entries were used, the page is empty and we can only restore eviction/restore
-         * or history store updates against empty row-store leaf pages, column-store modify attempts
-         * to allocate a zero-length array.
-         */
-        if (r->page->type != WT_PAGE_ROW_LEAF && chunk->entries == 0)
-            return (__wt_set_return(session, EBUSY));
+        if (chunk->entries == 0) {
+            /*
+             * XXX If no entries were used, the page is empty and we can only restore
+             * eviction/restore or history store updates against empty row-store leaf pages,
+             * column-store modify attempts to allocate a zero-length array.
+             */
+            if (r->page->type != WT_PAGE_ROW_LEAF)
+                return (__wt_set_return(session, EBUSY));
 
-        /* If we need to restore the page to memory, copy the disk image. */
-        if (multi->supd_restore)
+            /* If the row leaf page is empty and we need to restore the page to memory, copy the
+             * disk image. */
             goto copy_image;
-
-        WT_ASSERT_ALWAYS(session, chunk->entries > 0, "Trying to write an empty chunk");
+        }
     }
 
-    if (last_block)
-        WT_RET(__rec_build_delta(session, r));
-
     /* Write the disk image and get an address. */
-    WT_RET(__rec_write(session, compressed_image == NULL ? &chunk->image : compressed_image,
-      &chunk->block_meta, addr, &addr_size, &compressed_size, false, F_ISSET(r, WT_REC_CHECKPOINT),
-      compressed_image != NULL));
+    if (last_block)
+        WT_RET(__rec_build_delta(session, r, &build_delta));
+
+    if (build_delta) {
+        chunk->block_meta.is_delta = true;
+        WT_RET(__wt_blkcache_write(session, &r->delta, &chunk->block_meta, addr, &addr_size,
```

Comment:
Write delta.

Reply:
Thanks for working on this! I think it would actually make more sense to add a new `__wt_blkcache` function for the delta, as its behavior would be quite different from the regular write:
- Pass in the delta to PALI for write to the log server
- Update the entry in the block cache so that it would append the delta to the block's entry in the cache.

We could overload the write function to do this, but I think a new delta function would be cleaner. (We can continue to use the write function to write the full image.)

As for the metadata struct, PALI would need to know the some of the original metadata fields (once we add them) so that it can then pass in the corresponding info to the log service, e.g., for the various checkpoint IDs that it needs to know about. At the same time, we'll have to make some changes to the metadata fields (e.g., the number of deltas from last full image), but we'll have to decide at which level to do that - I _think_ the block manager level would be most appropriate, but we'll have to see. Maybe we can use some of the time in NY to discuss.